### ISSUE TITLE
[Troubleshoot]: PPO finetuning of openpi on Metaworld with dense rewards

### ISSUE THREAD

**Issue by brioAI** [RLinf/RLinf#515]
### Problem description

Hi,
I am interested in fine-tuning openpi on Metaworld 50 using Metaworld native dense reward.
Metaworld provides human engineered, geometric, dense reward for each timestep:
`raw_obs, _reward, _, _, infos = self.env.step(actions)`

1) is `_reward` the metaworld dense reward prebuilt [line 318 in metaworld_env.py](https://github.com/RLinf/RLinf/blob/779c3188b6576c9c93d43a860170f763d59e1a3d/rlinf/envs/metaworld/metaworld_env.py#L318C17-L318C69) 
`raw_obs, _reward, _, _, _ = self.env.step(zero_action)`

2) I used it as the reward in the function `_calc_step_reward` [line 446 in metaworld_env](https://github.com/RLinf/RLinf/blob/779c3188b6576c9c93d43a860170f763d59e1a3d/rlinf/envs/metaworld/metaworld_env.py#L446)
```
def _calc_step_reward(self, terminations, _reward=None):
  if self.use_dense_reward:
      return self.cfg.reward_coef * _reward
...remaining unchanged...
```
But the results of the fine-tuning of pi0.5 in Metaworld 50 is very poor:

<img width="1068" height="994" alt="Image" src="https://github.com/user-attachments/assets/bcd7d8ac-a286-4127-971c-2f897be3764c" />

My first guess was that the dense reward scaling was too large compared to the sparse reward (as indicated by the value loss dominating the policy loss), so I tried to downscale the dense reward by adding a downscaling coef of 0.005:

```
def _calc_step_reward(self, terminations, _reward=None):
  if self.use_dense_reward:
      return self.cfg.reward_coef * _reward * 0.005
...remaining unchanged...
```
Now the policy and value losses are of the same order of magnitude, but the fine-tuning still performs poorly. The training is still on-going but the eval results already feel off:

<img width="1068" height="846" alt="Image" src="https://github.com/user-attachments/assets/63830cfd-bbf8-4433-a65b-f63735db2e60" />

Have you tried using the dense reward instead of the sparse reward? Would you have suggestions to make the training work, like do you notice anything wrong in the curves?

Thanks a lot!

### Configuration YAML file

YAML config file:

```
defaults:
  - env/metaworld_50@env.train
  - env/metaworld_50@env.eval
  - model/pi0_5@actor.model
  - training_backend/fsdp@actor.fsdp_config
  - optional reward_model: null
  - override hydra/job_logging: stdout

hydra:
  run:
    dir: .
  output_subdir: null
  searchpath:
    - file://${oc.env:EMBODIED_PATH}/config/

cluster:
  num_nodes: 1
  component_placement:
    rollout: 0
    env: 1
    actor: 0-1

runner:
  task_type: embodied
  logger:
    log_path: "../results"
    project_name: rlinf
    experiment_name: "metaworld_50_ppo_openpi_pi05"
    logger_backends: ["tensorboard"] # wandb, swanlab

  max_epochs: 1000 # Global Epochs=Global Steps= one global step represents one full cycle of data collection → advantage computation → policy optimization
  max_steps: -1

  only_eval: False
  eval_policy_path: null # Optional: .pt file or None, if None, will use the checkpoint in rollout.model.model_path
  val_check_interval: 50 # Eval during training. Deactivate with a negative value e.g. -1. Activate with a positive value e.g. 100 - this determines how often (in epochs) evaluation runs
  save_interval: 50
  seq_length: 4096
  max_prompt_length: 50

  resume_dir: null

algorithm:
  normalize_advantages: True
  kl_penalty: kl  # how to estimate kl divergence: kl or kl_penalty
  group_size: 1
  reward_coef: 1.0

  rollout_epoch: 2 # Collects data 2 times, across total_num_envs environments
  eval_rollout_epoch: 8 # set eval_rollout_epoch > 0 when enable runner.only_eval or runner.val_check_interval > 0

  reward_type: chunk_level
  logprob_type: chunk_level
  entropy_type: token_level

  update_epoch: 4 # Trains the policy and value networks 4 times on the collected data
  adv_type: gae
  loss_type: actor_critic
  loss_agg_func: "token-mean" 
  kl_beta: 0.0
  entropy_bonus: 0
  clip_ratio_high: 0.2
  clip_ratio_low: 0.2
  clip_ratio_c: 3.0
  value_clip: 0.2
  huber_delta: 10.0

  # Advantage Estimation Parameters
  gamma: 0.99 # used for GAE
  gae_lambda: 0.95  # used for GAE

  filter_rewards: False
  rewards_lower_bound: 0.1
  rewards_upper_bound: 0.9
  # params for generation
  sampling_params:
    do_sample: True
    temperature_train: 1.0
    temperature_eval: 0.6
    top_k: 50
    top_p: 1.0
    repetition_penalty: 1.0
    add_BOS: False

  # length argument for autoregressive sampling
  # max length means max amount of tokens to generate
  length_params:
    max_new_token: null
    max_length: 1024
    min_length: 1

env:
  group_name: "EnvGroup"
  enable_offload: False

  # Override the default values in env/metaworld_50
  train:
    total_num_envs: 32
    use_dense_reward: true # added to use dense reward in a configurable way

  eval:
    total_num_envs: 64
    auto_reset: True
    ignore_terminations: True
    max_episode_steps: 100
    max_steps_per_rollout_epoch: 100
    group_size: 1
    use_fixed_reset_state_ids: True
    video_cfg:
      save_video: True
      video_base_dir: ${runner.logger.log_path}/video/eval

rollout:
  group_name: "RolloutGroup"
  generation_backend: "huggingface"
  recompute_logprobs: False
  unnorm_key: libero_10
  enable_offload: False
  pipeline_stage_num: 2

  model:
    model_path: "models/RLinf-Pi05-MetaWorld"
    precision: ${actor.model.precision}

actor:
  group_name: "ActorGroup"
  training_backend: "fsdp"
  micro_batch_size: 64
  global_batch_size: 256 # maximum : global batch size = micro batch size * actor num_processes.
  seed: 42
  enable_offload: False

  # Override the default values in model/pi0_5
  model:
    model_path: "models/RLinf-Pi05-MetaWorld"
    num_action_chunks: 5 # action chunks of pi0.5
    action_dim: 4 # metaworld action space
    add_value_head: True # Value head For PPO value estimation
    # openpi specific parameters
    openpi:
      config_name: "pi05_metaworld"
      num_images_in_input: 1
      value_after_vlm: True

  optim:
    lr: 1e-6
    value_lr: 1.0e-4
    adam_beta1: 0.9
    adam_beta2: 0.95
    adam_eps: 1.0e-05
    weight_decay: 0.01
    clip_grad: 1.0
    critic_warmup_steps: 0
    total_training_steps: ${runner.max_epochs}
    lr_warmup_steps: -1
    warmup_style: cosine
    num_cycles: 0.5

  # Override the default values in training_backend/fsdp
  fsdp_config:
    strategy: "fsdp"
    sharding_strategy: "no_shard"
    gradient_checkpointing: False # for openpi, gradient checkpointing is not supported, please do not change this value
    mixed_precision:
      param_dtype: ${actor.model.precision}
      reduce_dtype: ${actor.model.precision}
      buffer_dtype: ${actor.model.precision}

reward:
  use_reward_model: False

critic:
  use_critic_model: False
```

### Log file

/

### Environment

python -V
uv pip list
nvidia-smi
Python 3.12.12
Package                 Version
----------------------- ---------
archspec                0.2.5
boltons                 25.0.0
brotli                  1.1.0
certifi                 2025.10.5
cffi                    2.0.0
charset-normalizer      3.4.4
colorama                0.4.6
conda                   25.9.1
conda-libmamba-solver   25.4.0
conda-package-handling  2.4.0
conda-package-streaming 0.12.0
distro                  1.9.0
frozendict              2.4.6
h2                      4.3.0
hpack                   4.1.0
hyperframe              6.1.0
idna                    3.11
jsonpatch               1.33
jsonpointer             3.0.0
libmambapy              2.3.3
menuinst                2.4.1
packaging               25.0
pip                     25.2
platformdirs            4.5.0
pluggy                  1.6.0
pycosat                 0.6.6
pycparser               2.22
pysocks                 1.7.1
requests                2.32.5
ruamel-yaml             0.18.16
ruamel-yaml-clib        0.2.14
setuptools              80.9.0
tqdm                    4.67.1
truststore              0.10.3
urllib3                 2.5.0
wheel                   0.45.1
zstandard               0.25.0

NVIDIA H100 80GB HBM3

### Before submitting a new issue...

- [x] Have you checked relevant issues, FAQs (https://rlinf.readthedocs.io/en/latest/rst_source/faq.html), or asked the chatbot at the top right corner of the [documentation page](https://rlinf.readthedocs.io).

**Comment by chenkang455** [comment]
Hi, @brioAI I haven't tested the dense reward function yet, but the multi-stage reward we’re using in CALVIN has been effective. Could you provide the explained variance curve? It would help me assess if the critic is well-trained. Also, I suggest trying pi0 on Meta-World, as it generally outperforms pi0.5.

I notice that the approx-kl is relatively low. Can you enlarge the lr for another try?

**Comment by brioAI** [comment]
Hi @chenkang455,
Sure here is the explained variance curve:

<img width="1235" height="359" alt="Image" src="https://github.com/user-attachments/assets/eae0c4b4-d8aa-4319-aff8-580bfe0bf9af" />

Here are the other train curves if needed:

<img width="1235" height="1053" alt="Image" src="https://github.com/user-attachments/assets/542e6a27-7f3f-4a3f-9a39-a198ae0bf7ce" />

Do you see anything suspicious in these additional curves? Which value would you judge as normal for the approx-kl? Here it fluctuates between 0.0015 and 5e-4


I will try enlarging the learning rate for another try, should I keep the dense reward downscaled like my 2nd try too? By enlarging do you mean going from 1e-6 to 3e-6?

**Comment by chenkang455** [comment]
I noticed that the learning rate in your configuration (1e-6) is inconsistent with our recommended setting (5e-6). Could you try updating it to 5e-6? Additionally, please monitor the approx-kl; it typically stays within the 0.005 to 0.02 range.

**Comment by brioAI** [comment]
Hi @chenkang455 thanks for the reply, I downgraded the learning rate following your previous post:
https://github.com/RLinf/RLinf/issues/470#issuecomment-3680721478
as I use a lower resource config than the one you use

**Comment by chenkang455** [comment]
Get it, but I think for your case, approx-kl seems low. You can try a larger lr or a higher update-epoch.

<img width="1504" height="816" alt="Image" src="https://github.com/user-attachments/assets/590a4b9a-26cc-4882-bcd1-95f8301c0e2b" />

**Comment by brioAI** [comment]
Thanks for the recommendation, I will try the higher lr: 5e-6, with the same config as before (train.total_num_envs: 32, actor.global_batch_size: 256, actor.micro_batch_size: 64, algorithm.rollout_epoch: 2, algorithm.update_epoch: 4)
I will let you know.

Did you do a typo in your message by saying "please monitor the approx-kl; it typically stays within the 0.05 to 0.2 range." ? Your plot shows an approx-kl between 0.005 and 0.015 which is one magnitude lower than your message.

**Comment by chenkang455** [comment]
@brioAI Yes, sry for that, it should be 0.005 - 0.02. Looking forward to your result!

**Comment by brioAI** [comment]
The approx_kl is now in the correct range, but the training does not go well at all :-(

<img width="1156" height="1078" alt="Image" src="https://github.com/user-attachments/assets/bfcf1fcb-bf9b-49c6-b820-64a746bc4ef9" />

config
```
algorithm:
  rollout_epoch: 2
  eval_rollout_epoch: 8
  update_epoch: 4
  train:
    total_num_envs: 32

  optim:
    lr: 5e-6
    value_lr: 1.0e-4

actor:
  micro_batch_size: 64
  global_batch_size: 256
```

Maybe training only on the dense reward is not good and an hybrid approach dense-parse would be better, where we give a bonus at the end of a successful completion.

**Comment by chenkang455** [comment]
Thanks for providing it. I agree that combining both rewards is likely more effective. We could test different weight distributions—for instance, starting with a (0, 1.0) split for dense and sparse rewards versus a balanced (0.5, 0.5) approach—to see which yields better performance.

**Comment by brioAI** [comment]
Interesting, I will try and let you know.

**Comment by chenkang455** [comment]
> Interesting, I will try and let you know. 

Looking forward to it! Besides, I recommend to conduct experiment on pi0, which performs better in metaworld.
